core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -1632,282 +1665,265 @@ object ZSTM {
+      def getOrElseUpdate(key: TRef[?], entry: => Entry): Entry =
+        map.getOrElseUpdate(key, entry)
 
-    /**
-     * Allocates memory for the journal, if it is null, otherwise just clears
-     * it.
-     */
-    def allocJournal(journal: Journal): Journal =
-      if (journal eq null) new MutableMap[TRef[_], Entry](DefaultJournalSize)
-      else {
-        journal.clear()
-        journal
-      }
+      def keys: collection.Set[TRef[?]] = map.keySet
 
-    /**
-     * Determines if the journal is valid.
-     */
-    def isValid(journal: Journal): Boolean = {
-      var valid = true
-      val it    = journal.entrySet.iterator
-      while (valid && it.hasNext) valid = it.next.getValue.isValid
-      valid
-    }
+      // -- Transactional API --
 
-    /**
-     * Analyzes the journal, determining whether it is valid and whether it is
-     * read only in a single pass. Note that information on whether the journal
-     * is read only will only be accurate if the journal is valid, due to
-     * short-circuiting that occurs on an invalid journal.
-     */
-    def analyzeJournal(journal: Journal): JournalAnalysis = {
-      var result = JournalAnalysis.ReadOnly: JournalAnalysis
-      val it     = journal.entrySet.iterator
-      while ((result ne JournalAnalysis.Invalid) && it.hasNext) {
-        val value = it.next.getValue
-        if (value.isInvalid) result = JournalAnalysis.Invalid
-        else if (value.isChanged) result = JournalAnalysis.ReadWrite
+      /**
+       * Analyzes the journal, determining whether it is valid and whether it is
+       * read only in a single pass. Note that information on whether the
+       * journal is read only will only be accurate if the journal is valid, due
+       * to short-circuiting that occurs on an invalid journal.
+       *
+       * In the case that there is only a single entry in the journal, we can
+       * further shortcut and attempt to commit the transaction in a single pass
+       * if the `attemptCommit` parameter is set to `true`
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[internal] def analyze(attemptCommit: Boolean): JournalAnalysis =
+        if (map.size == 1) analyzeJournal1(map.head._2, attemptCommit)
+        else analyzeJournalN()
+
+      private[this] def analyzeJournal1(value: Entry, attemptCommit: Boolean): JournalAnalysis =
+        if (attemptCommit) {
+          if (value.attemptCommit()) JournalAnalysis.Committed
+          else JournalAnalysis.Invalid
+        } else if (value.isInvalid) JournalAnalysis.Invalid
+        else if (value.isChanged) JournalAnalysis.ReadWrite
+        else JournalAnalysis.ReadOnly
+
+      private[this] def analyzeJournalN(): JournalAnalysis = {
+        var changed = false
+
+        val it = map.valuesIterator
+        while (it.hasNext) {
+          val value = it.next()
+          if (value.isInvalid) return JournalAnalysis.Invalid
```

Comment:
haven't seen that in a while 😄 non-local returns are deprecated in scala 3, it uses exceptions under the hood so the performance is not optimal. better use an `isValid` var instead.

Reply:
This is not a non-local return :) Non-local returns are when the `return` keyword is contained within a Function. When they're used in a method directly they're just normal return statements.

I explained this in a bit more detail in this PR https://github.com/zio/zio/pull/9016#discussion_r1680690960 previously if you're interested

Reply:
oh, interesting that the semantics differs based on the use site, thanks for the details!